packages/component-base/test/data-provider-controller.test.js

```diff
@@ -0,0 +1,226 @@
+
+  describe('with size', () => {
+    beforeEach(() => {
+      controller = new DataProviderController(host, {
+        size: 500,
+        pageSize: 50,
+        isExpanded,
+        dataProvider: (_params, callback) => callback([], 0),
+      });
+    });
+
+    it('should have size', () => {
+      expect(controller.size).to.equal(500);
+    });
+
+    it('should have rootCache size', () => {
+      expect(controller.rootCache.size).to.equal(500);
+    });
+  });
+
+  describe('with dataProviderParams', () => {
+    let dataProviderParams;
+
+    beforeEach(() => {
+      dataProviderParams = () => ({ filter: 'bar' });
+
+      controller = new DataProviderController(host, {
+        pageSize: 50,
+        isExpanded,
+        dataProvider: (_params, callback) => callback([], 0),
+        dataProviderParams,
+      });
+    });
+
+    it('should have dataProviderParams', () => {
+      expect(controller.dataProviderParams).to.equal(dataProviderParams);
+    });
+  });
+
+  describe('clearCache', () => {
+    beforeEach(() => {
+      controller = new DataProviderController(host, {
+        size: 500,
+        pageSize: 50,
+        isExpanded,
+        dataProvider: (_params, callback) => callback([], 0),
+      });
+    });
+
+    it('should create a new cache', () => {
+      const { rootCache } = controller;
+      controller.clearCache();
+      expect(controller.rootCache).to.not.equal(rootCache);
+    });
+
+    it('should set size for the new cache', () => {
+      controller.clearCache();
+      expect(controller.rootCache.size).to.equal(500);
+    });
+
+    it('should set pageSize for the new cache', () => {
+      controller.clearCache();
+      expect(controller.rootCache.pageSize).to.equal(50);
+    });
+  });
+
+  describe('setSize', () => {
+    beforeEach(() => {
+      controller = new DataProviderController(host, {
+        pageSize: 50,
+        isExpanded,
+        dataProvider: (_params, callback) => callback([], 0),
+      });
+    });
+
+    it('should update size', () => {
+      controller.setSize(100);
+      expect(controller.size).to.equal(100);
+    });
+
+    it('should update rootCache size', () => {
+      controller.setSize(100);
+      expect(controller.rootCache.size).to.equal(100);
+    });
+
+    it('should recalculate flatSize', () => {
+      controller.setSize(100);
+      expect(controller.flatSize).to.equal(100);
+    });
+  });
+
+  describe('setPageSize', () => {
+    beforeEach(() => {
+      controller = new DataProviderController(host, {
+        pageSize: 50,
+        isExpanded,
+        dataProvider: (_params, callback) => callback([], 0),
+      });
+    });
+
+    it('should update pageSize', () => {
+      controller.setPageSize(10);
+      expect(controller.pageSize).to.equal(10);
+    });
+
+    it('should update rootCache pageSize', () => {
+      controller.setPageSize(10);
+      expect(controller.rootCache.pageSize).to.equal(10);
+    });
+
+    it('should clear the cache', () => {
```

Comment:
It's actually a bit hard to define what should happen in this particular case (pageSize).

- The root cache has the correct `pageSize` (already tested)
- The root cache has no pending requests (should they be completely cleared or remove only those that ended up out of the valid range?)
- The root cache has no sub-caches OR pageSize is propagated to all the descendant caches.

Reply:
Discussed this internally. Not an issue now, especially since we have a lot of tests from a use-case perspective (like grid tests now, and combo-box tests later). We can improve these couple of tests later if necessary.